muxers/mplex/src/codec.rs

```diff
@@ -114,29 +113,27 @@ impl Decoder for Codec {
                     match self.varint_decoder.decode(src)? {
                         Some(len) => {
                             if len as usize > MAX_FRAME_SIZE {
-                                return Err(IoErrorKind::InvalidData.into());
+                                let msg = format!("Mplex frame length {} exceeds maximum", len);
+                                return Err(IoError::new(IoErrorKind::InvalidData, msg));
                             }
 
-                            self.decoder_state = CodecDecodeState::HasHeaderAndLen(header, len as usize, BytesMut::with_capacity(len as usize));
+                            self.decoder_state = CodecDecodeState::HasHeaderAndLen(header, len as usize);
                         },
                         None => {
                             self.decoder_state = CodecDecodeState::HasHeader(header);
                             return Ok(None);
                         },
                     }
                 },
-                CodecDecodeState::HasHeaderAndLen(header, len, mut buf) => {
-                    debug_assert!(len == 0 || buf.len() < len);
-                    let to_transfer = cmp::min(src.len(), len - buf.len());
-
-                    buf.put(src.split_to(to_transfer));    // TODO: more optimal?
-
-                    if buf.len() < len {
-                        self.decoder_state = CodecDecodeState::HasHeaderAndLen(header, len, buf);
+                CodecDecodeState::HasHeaderAndLen(header, len) => {
+                    if src.len() < len {
+                        self.decoder_state = CodecDecodeState::HasHeaderAndLen(header, len);
+                        let to_reserve = len - src.len();
+                        src.reserve(to_reserve);
```

Comment:
I didn't do that because the implementation already does it: https://docs.rs/bytes/0.4.9/src/bytes/bytes.rs.html#2195
Either tokio's implementation is too defensive, or its documentation is outdated, I don't know which.

Reply:
Looking at https://docs.rs/bytes/0.4.9/src/bytes/bytes.rs.html#2191 I think that lines 2192 to 2199 are equivalent to checking via `remaining_mut` first so there is no value in duplicating that logic.